python-sdk/rebuff/rebuff_pythonsdk.py

```diff
@@ -0,0 +1,228 @@
+    ) -> Tuple[Any, str]:
+        """
+        Adds a canary word to the given prompt which we will use to detect leakage.
+
+        Args:
+            prompt (Any): The prompt to add the canary word to.
+            canary_word (Optional[str], optional): The canary word to add. If not provided, a random canary word will be
+             generated. Defaults to None.
+            canary_format (str, optional): The format in which the canary word should be added.
+            Defaults to "<!-- {canary_word} -->".
+
+        Returns:
+            Tuple[Any, str]: A tuple containing the modified prompt with the canary word and the canary word itself.
+        """
+
+        # Generate a canary word if not provided
+        if canary_word is None:
+            canary_word = self.generate_canary_word()
+
+        # Embed the canary word in the specified format
+        canary_comment = canary_format.format(canary_word=canary_word)
+        if isinstance(prompt, str):
+            prompt_with_canary: str = canary_comment + "\n" + prompt
+            return prompt_with_canary, canary_word
+
+        try:
+            if isinstance(prompt, langchain.prompts.PromptTemplate):
+                prompt.template = canary_comment + "\n" + prompt.template
+                return prompt, canary_word
+
+        except Exception as error:
+            raise Exception(error)
+
+    def is_canary_word_leaked(
+        self,
+        user_input: str,
+        completion: str,
+        canary_word: str,
+        log_outcome: bool = True,
+    ) -> bool:
+        """
+        Checks if the canary word is leaked in the completion.
+
+        Args:
+            user_input (str): The user input.
+            completion (str): The completion generated by the AI.
+            canary_word (str): The canary word to check for leakage.
+            log_outcome (bool, optional): Whether to log the outcome of the leakage check. Defaults to True.
+
+        Returns:
+            bool: True if the canary word is leaked, False otherwise.
+        """
+        if canary_word in completion:
+            if log_outcome:
+                self.log_leakage(user_input, completion, canary_word)
+            return True
+        return False
+
+    def log_leakage(self, user_input: str, completion: str, canary_word: str) -> None:
+        """
+        Logs the leakage of a canary word.
+
+        Args:
+            user_input (str): The user input.
+            completion (str): The completion generated by the AI.
+            canary_word (str): The leaked canary word.
+        """
+
+        vector_store = init_pinecone(
+            self.pinecone_environment,
+            self.pinecone_apikey,
+            self.pinecone_index,
+            self.openai_apikey,
+        )
+
+        vector_store._text_key = "input"
+        vector_store.add_texts(
+            [user_input],
+            metadatas=[{"completion": completion, "canary_word": canary_word}],
```

Comment:
I think we should make sure the metadata has the same keys whether it's coming from JS or Python (in JS, the key is `"canaryWord"` instead of `"canary_word"`). I'd actually recommend changing the JS to match what you have in Python here since snake case is more common for names of data fields than camel case.

Reply:
Yes, I think we would want the metadata keys to be same. Though I have not updated the JS code yet since I am not sure how [metadata keys](https://github.com/protectai/rebuff/blob/main/javascript-sdk/src/sdk.ts#L152C50-L152C60) are being setup in JS.

Reply:
In TypeScript, when you initialize a `Record` (same thing as a `dict` in Python) using this syntax:
```ts
const metadata = { completion, canaryWord };
```
the variable name becomes the key and the variable value becomes the value. It's the same as doing:
```ts
const metadata = { "completion": completion, "canaryWord": canaryWord };
```
So, you could change the line you linked to:
```ts
        this.logLeakage(userInput, { completion, "canary_word": canaryWord });
```
If you like, I can do this change in a separate PR.

Reply:
Thank you for the detailed explanation. I have made the change as you suggested, though would appreciate if you can review it again please.

Reply:
Yep, that change looks good. :+1: